### PR TITLE
feat: support gRPC for Rewrite objects

### PR DESCRIPTION
**Pull request by coryan**
We need to support gRPC-created Rewrite objects.

Part of the work for #192

### CODE REVIEW

File: gcs/rewrite.py [hunk 1]
```diff
@@ -56,25 +60,64 @@ def init_rest(
...
-        token = hashlib.sha256(
-            (
-                "%s/o/%s/rewriteTo/b/%s/o/%s"
-                % (src_bucket_name, src_object_name, dst_bucket_name, dst_object_name)
-            ).encode("utf-8")
-        ).hexdigest()
+
+    @classmethod
+    def init_grpc(cls, request, context):
+        if not request.source_bucket.startswith("projects/_/buckets/"):
```
Comment (tritone) [inline gcs/rewrite.py:78]: Does JSON have this validation logic? If not should it be shared?

**Review by tritone — COMMENTED** [review]
Generally looks good, one question.

File: gcs/rewrite.py (hunk 1 above)
Comment (coryan) [reply gcs/rewrite.py:78]: JSON bucket names are not prefixed by `projects/_/buckets/` so the validation cannot be identical. I assume you are asking about the other checks (non-empty and so forth).  I think the bucket names being non-empty is validated implicitly by the Flask route decorators, e.g.:

https://github.com/googleapis/storage-testbench/blob/c1d3ec4e96a519e1b1f1d6077eaf96575678fb5c/testbench/rest_server.py#L595-L600

Comment (tritone) [reply gcs/rewrite.py:78]: Yep that's correct-- sounds good, thanks!

**Review by tritone — APPROVED** [review]

**Pull request merged**